[webview_flutter_android][webview_flutter_wkwebview] Adds support for `setOnScrollPositionChange` for webview_flutter platform implementations

TheVinhLuong opened the pull request:
Adds iOS and Android implementation for content offset listener

This PR is part of a series of PRs that aim to close https://github.com/flutter/flutter/issues/31027.

The PR that contains all changes can be found at https://github.com/flutter/packages/pull/3444.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

TheVinhLuong commented:
Hi @bparrishMines, could you please review and merge it?

hellohuanlin commented on packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFScrollViewDelegateHostApi.m:
```diff
@@ -0,0 +1,90 @@
+// references.
+@property(nonatomic, weak) id<FlutterBinaryMessenger> binaryMessenger;
+// InstanceManager must be weak to prevent a circular reference with the object it stores.
+@property(nonatomic, weak) FWFInstanceManager *instanceManager;
```
can you explain more about this circular reference?

TheVinhLuong replied:
Sorry, but that comment is actually not come up by me, you can see that comment throughout the code base wherever the property of `FWFInstanceManager ` is declared, I just followed the convention. 

However, from my understanding, it is a safety measure to prevent a memory leak in case the instance of `FWFInstanceManager` is used to store the object that contains it (circular relation).

bparrishMines replied:
@hellohuanlin This is the pattern for all the api implementations in the plugin. The full reasoning is that the `ScrollViewDelegate` instance contains an instance of the `ScrollViewDelegateFlutterApi`. And the `ScrollViewDelegateFlutterApi` contains an instance of the `FWFInstanceManager`. And the `FWFInstanceManager` contains an instance of the `ScrollViewDelegate`.

`ScrollViewDelegate` -> `FlutterApi` -> `InstanceManager` -> `ScrollViewDelegate`. 

I'm currently working on https://github.com/flutter/flutter/issues/134777 to eventually make this a part of generated code and not copy pasted.

hellohuanlin commented on packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFScrollViewDelegateHostApi.m:
```diff
@@ -0,0 +1,90 @@
+                          uiScrollView:(UIScrollView *)scrollView
+                            completion:(void (^)(FlutterError *_Nullable))completion {
+  [self scrollViewDidScrollWithIdentifier:[self identifierForDelegate:instance]
+                   uiScrollViewIdentifier:[self.instanceManager
```
if instanceManager is weak, do you want to deal with nil case?

TheVinhLuong replied:
I don't think `instanceManager` can be nil here. Again, I just followed the pattern of other host API was implemented before me.

The reason why I think `instanceManager` couldn't be nil is because it is strongly referenced by the plugin [here](https://github.com/flutter/packages/commit/313ef61bc1b3b6b41e38573641d46ab49afe40ea#diff-97d58ad121730325019c0434ec1c12c2a47c62e07c6bb09a54e48a619d695601R107)

bparrishMines replied:
`instanceManager` can be null in an add-to-app scenario. However, Objective-c will just return `nil` and not cause a crash. This is the desired behavior since the plugin holds the only strong reference and we want this to be null when that changes. I will make this clearer when https://github.com/flutter/flutter/issues/134777 is implemented in this language.

hellohuanlin commented on packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFScrollViewDelegateHostApi.m:
```diff
@@ -0,0 +1,90 @@
+  [self.scrollViewDelegateAPI scrollViewDidScrollForDelegate:self
+                                                uiScrollView:scrollView
+                                                  completion:^(FlutterError *error) {
+                                                    NSAssert(!error, @"%@", error);
```
should we pipe the error to flutter side?

TheVinhLuong replied:
As I understand, the `completion` callback here is for the native side to handle the error pipe from the flutter side, so I don't think it is logical to passing it back to the flutter side. Besides, every callback that was implemented before me is doing the same, I just followed the convention.

hellohuanlin commented on packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFScrollViewDelegateHostApi.m:
```diff
@@ -0,0 +1,90 @@
+  if (self) {
+    _scrollViewDelegateAPI =
+        [[FWFScrollViewDelegateFlutterApiImpl alloc] initWithBinaryMessenger:binaryMessenger
+                                                             instanceManager:instanceManager];
```
this seems to be just a 1-line wrapper. can `FWFScrollViewDelegate` itself conform to `FWFScrollViewDelegateFlutterApi`?

TheVinhLuong replied:
I believe it's possible, but I adhered to the conventions established by the previously implemented API, such as the FWFUIDelegate.

bparrishMines replied:
The reason this pattern doesn't have the wrapped class implement their own FlutterApi is because it wouldn't work well with inheritance. For example if we had

```dart
class A { }
class B implements A { }

interface AFlutterApi { } 
interface BFlutterApi { }
```

`class B` would need access to an implementation of `AFlutterApi` and `BFlutterApi`. So, there wouldn't be an easy way to do this if we did `class A implements AFlutterApi`. 

Note that this will also be solved by https://github.com/flutter/flutter/issues/134777 by handling inherited flutter api automatically.

hellohuanlin reviewed (commented):
reviewed ios part

TheVinhLuong commented:
@hellohuanlin Thank you for your review! By the way, this PR of mine is extracted from my PR [here](https://github.com/flutter/flutter/issues/31027), which has already been reviewed and approved by 2 people. That is why, in my opinion, it is a waste of effort to be thoroughly reviewed again. What I think should be reviewed here is that are there any difference between the code change here compared to the PR that contains a full changes (including platform interface,...)

bparrishMines reviewed (approved):
LGTM 

@hellohuanlin This code was already reviewed in https://github.com/flutter/packages/pull/3444 and follows the design from https://docs.google.com/document/d/1wXB1zNzYhd2SxCu1_BK3qmNWRhonTB6qdv4erdtBQqo/edit?usp=sharing&resourcekey=0-WOBqqOKiO9SARnziBg28pg. Let me know if you have questions about it.

cc @stuartmorgan

bparrishMines commented:
I merged in the latest changes from the main branch, but this is still blocked by an ifra error failing the tests.

TheVinhLuong commented:
I just changed the unit test a bit.

bparrishMines commented:
@stuartmorgan This is ready for a secondary review. 

@hellohuanlin You can secondary review the iOS portion too if you would like. 

I will handle the version bump once this is done.

hellohuanlin commented on packages/webview_flutter/webview_flutter_wkwebview/example/ios/RunnerTests/FWFScrollViewDelegateHostApiTests.m:
```diff
@@ -0,0 +1,82 @@
+  FWFScrollViewDelegateFlutterApiImpl *flutterAPI = [[FWFScrollViewDelegateFlutterApiImpl alloc]
+      initWithBinaryMessenger:OCMProtocolMock(@protocol(FlutterBinaryMessenger))
+              instanceManager:instanceManager];
+  return OCMPartialMock(flutterAPI);
```
i'd just inline these 2 helpers.

stuartmorgan commented on packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart:
```diff
@@ -849,6 +853,35 @@ abstract class NSUrlFlutterApi {
+@FlutterApi()
+abstract class UIScrollViewDelegateFlutterApi {
+  @ObjCSelector(
+    'scrollViewDidScrollWithIdentifier:uiScrollViewIdentifier:x:y:',
```
Nit: ObjC naming should be `...:UIScrollViewIdentifier:x:y:`. Capitalizing acronyms trumps first letter being lower case for camel case in Obj-C naming conventions.

TheVinhLuong replied:
Lemme fix that.

stuartmorgan reviewed (approved):
LGTM with nit

jmagman commented:
@TheVinhLuong can you fix up the conflicts, including the CHANGELOG and pubspec version?  Once @stuartmorgan's `UIScrollViewIdentifier` issue is addressed this will be ready to merge!

TheVinhLuong commented:
> @TheVinhLuong can you fix up the conflicts, including the CHANGELOG and pubspec version? Once @stuartmorgan's `UIScrollViewIdentifier` issue is addressed this will be ready to merge!

I fixed up the code conflict. Also, the `UIScrollViewIdentifier` issue also addressed.
